doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+However, this can lead to less efficient code, if the second argument of `f` is used more than once.
+For that reason, it may be a good idea to provide both a strict version and a non-strict version of `foldr`.
+
+Some strict languages have special syntaxes to mark a function lazy in a particular argument, e.g., Scala has the `=>` syntax.
+The following function is strict in the first argument and lazy in the second argument:
+
+```scala
+def foo(a: Int, b: => Int) { ... }
+```
+
+`nonstrict` can be used to achieve a similar effect, except that `nonstrict` is used at the call sites as opposed to function definitions.
+
+#### `foldr` and `nonstrict` in Other Strict Languages
+
+Scala and OCaml are both strict by default.
+In both languages, their version of `foldr` for lists in the respective standard library ([Scala](https://www.scala-lang.org/api/2.13.3/scala/collection/immutable/List.html#foldRight[B](z:B)(op:(A,B)=%3EB):B),[OCaml](https://v2.ocaml.org/api/List.html)) is strict and doesn't short circuit.
+Functions that should short-circuit, such as `all` and `any`, do short-circuit, and unsurprisingly they are not implemented in terms of `foldr`.
+
+Scala's [cats library](https://github.com/typelevel/cats) provides a `Foldable` typeclass, and a short-circuiting `foldRight`.
+In order to do so, cats' `foldRight` has a different type than the one in the standard library: the type of the accumulator is `Eval b` rather than `b`.
+Our `nonstrict` function is similar to `Eval.always`, except that `nonstrict`'s type is `a -> a` while the type of `Eval.always` is `a -> Eval a` - because it is a library, not a plugin.
+Incidentally, there's a simpler way to make `foldRight` lazy using Scala's built-in syntax for lazy arguments, but it is not stack safe.
+Interestingly, the `Foldable.forall` and `Foldable.exists` functions in cats are implemented by [converting the foldable structure to list first](https://github.com/typelevel/cats/blob/896f00cd689af20676556d1b57791d7ade5cfa04/core/src/main/scala/cats/Foldable.scala#L677-L686), rather than using `foldRight`.
+
+We can potentially introduce something similar to `Eval` in the Plutus Tx standard library.
+The advantage of `Eval` is that it can be provided entirely in a library, without special compiler support. However, this is also what leads to its disadvantage: `Eval` is more heavyweight to use, since it changes the types of many functions (such as `foldr`), while `nonstrict` has the same type as the identity function, which is possible because it is a magical function that the compiler treats specially.
+
+The takeaway is this: the standard libraries usually do the simple things, e.g., providing a monomorphic, non-short-circuiting version of `foldr` for lists, as well as functions like `all`, `any`, also monomorphically for lists.
+The polymorphic versions of those functions, if exist, are usually provided by third-party libraries like cats.
```

Comment:
I think Scala isn't a good example here: the standard library is sort of half-heartedly functional, one of the distinguishing features of `cats` is to try and be much more serious about Haskell-inspired functional programming.

Reply:
Yes, but AFAIK OCaml's `foldr` in its stdlib is also strict. It seems to make sense for a strict language to provide a strict `foldr`.

Also I'm not aware of a language that is perfectly analogous to Plutus Tx such that we can just copy what it does. E.g., for OCaml the choice is strict vs. lazy, but for Plutus Tx it is strict vs non-strict, which are totally different dynamics.